### ISSUE TITLE
Warp file from MICCAI version

### ISSUE THREAD

**Issue by vmadge** [voxelmorph/voxelmorph#81]
Hello,

This is a follow-up to a previous (closed) issue regarding the warp from the MICCAI version of the code. I know the warp file from the MICCAI version is the concatenated parameters of the flow field (mu and log_sigma), but I was hoping to obtain the warp transform to apply to manual labels, bring them into the common space, and compare the labels of the volumes post-registration. I was wondering if there was a way to obtain the same transform that was applied to the moving image?

Thank you!

**Comment by adalca** [comment]
Sure -- the `miccai2018_net` model has a layer called `diffflow` -- this is the layer that is applied to the first image to be warped to the second (it's the integral of the velocity field sample). You can therefore create a new model from your current model: something like `new_model = keras.models.Model(model.inputs, model.get_layer('diffflow').outputs)` . Now the new model will output that deformation field. 

Also, note that you can continue commenting on closed issues -- they are only closed if we think that issue has been solved, but if you think you have a related question you can keep asking.

**Comment by vmadge** [comment]
Hello,

I am following up again on this issue of the deformation field output from VoxelMorph. I need the warp from VoxelMorph in order to measure the recovery error from my images, and compare the recovery error from ANTs SyN and a few other non-linear techniques. I want to make sure that I am best representing VoxelMorph results. 

The warp was obtained as suggested, and thus outputs as an npz file. I take this file and convert to a minc grid (with correct orientation) + corresponding xfm file. I then resample the labels with the warp externally from VoxelMorph. However, I noticed that the recovered label does not look like the recovered output from VoxelMorph. For example:

![image](https://user-images.githubusercontent.com/12600191/82130452-6e078b00-9799-11ea-862a-3e7f5b7dbdef.png)

I believe the issue is the label is resampled differently than how VoxelMorph resamples the output image. However, there is no way to access the resampler within VoxelMorph. I was hoping you could help determine the best way to extract the warp? I have thought of a few methods to approach this problem:

- Use the resampler externally? 
- Pass the label through VoxelMorph such that it is resampled by the exact same warp as it's recovered output? 
- Create a 3D grid internally such that a transform can be created externally from x-x', y-y', z-z'?

If you have the time to help I would sincerely appreciate it.

Thank you.

**Comment by adalca** [comment]
Hi @vmadge ,

Apologies, I'm not quite sure I understand the problem (I should also mention that I dont know what system you are using and what your process is -- I assume a "minc grid" is ANTs terminology?). 

You said "there is no way to access the resampler within VoxelMorph" -- what do you mean by this? All of VoxelMorph is open source and you can find the resampler we use through `SpatialTransformer`, which uses `neuron.utils.transform`. It is probably true taht you cannot use the voxelmorph fields with another tool that makes different assumptions.

I think we should start with this: what is it that you are trying to achieve the VoxelMorph does not provide?

**Comment by vmadge** [comment]
Thanks for your quick reply @adalca. I am trying to resample the label corresponding to the image using the same warp which resamples the recovered output from VoxelMorph. But it is important also to obtain that very warp which resamples the recovered output so I can compute the recovery error. 

Your reply indicates that I can do this via external access to SpatialTransformer? Exactly what command would I need to do so? Otherwise is there another way to obtain the same warp? 

Thanks.

**Comment by adalca** [comment]
sorry, what do you mean by "same warp"? 

Given two images, VoxelMorph gives you to warp between them (e.g. warp from src to target, so the vectors are in the space of the target).

With this warp you can warp anything from the source space. You can warp the original image, or a label map, or whatever you want. 

Which version of voxelmorph you are using or trained? We are currently in a transition to a new structure of the code, but if you are using the `master` branch and the `cvpr` model for example, then I recommend you look at https://github.com/voxelmorph/voxelmorph/blob/master/src/test.py to see how we registered two volumes, then used the resulting field to warp a label map.

**Comment by vmadge** [comment]
"same warp" refers to the warp between two images (fixed + moving). I would like to use this warp to resample the moving label map so it looks like the moved image. 

I am using the miccai2018 pretrained version.

The resulting field which I obtain from VoxelMorph does not resample the label map to look like the moved image outputted from voxelmorph. 

I would like to find a way to resample using VoxelMorph. Can I do this by running Spatial Transformer separately from VoxelMorph?

Thanks.

**Comment by adalca** [comment]
@vmadge can you please take  a look through https://github.com/voxelmorph/voxelmorph/blob/master/src/test_miccai2018.py

I believe it does what you want. Specifically, the `nn transfer model` we create there takes in a warp and label map and it warps it using nearest neighbor.
